### ISSUE TITLE
Slots not checked if `Self` annotation is present

### ISSUE THREAD

**Issue by alwaysmpe** [python/mypy#19331]
**Bug Report**

When using a dataclass with `slots=True` or `__slots__`, if an argument to a function is annotated with `Self` (not necessarily the `self` argument), then assigning to an attribute not annotated in slots doesn't raise an error.

**To Reproduce**
Type check the below ([playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=87ddfc2fb7de73512b2d91a6c8a616f9)). 
```python
from dataclasses import dataclass
from typing import Self

@dataclass(slots=True)
class Exp:
    x: int
    def foo(self) -> None:
        self.y = 1
    def bar(self, arg: Self) -> None:
        self.y = 1
    def baz(self: Self) -> None:
        self.y = 1

class Exp2:
    __slots__ = 'x',
    x: int
    def foo(self) -> None:
        self.y = 1
    def bar(self, arg: Self) -> None:
        self.y = 1
    def baz(self: Self) -> None:
        self.y = 1
```

**Expected Behavior**

All 3 invalid asignments to `y` would raise an error, dataclass or not.

**Actual Behavior**

Only the `foo` function raises an error for assigning to an attribute not in slots.
```text
main.py:8: error: Trying to assign name "y" that is not in "__slots__" of type "__main__.Exp"  [misc]
main.py:18: error: Trying to assign name "y" that is not in "__slots__" of type "__main__.Exp2"  [misc]
Found 2 errors in 1 file (checked 1 source file)
```
**Your Environment**

- Mypy version used: 1.16.1
- Python version used: 3.12

**Comment by alwaysmpe** [comment]
I think the problem is in `check_assignment_to_slots`:
https://github.com/python/mypy/blob/16e99de5376464beaa2cf086c1cd3dc5d26a791a/mypy/checker.py#L3701-L3702
as `Self` is a `TypeVarType`, need to add something like
```python
if isinstance(inst, TypeVarType):
    inst = inst.upper_bound
```
Which results in the correct error locally but I've not run the rest of the tests and it's late right now

**Comment by sterliakov** [comment]
Ough, sorry, I didn't mean to take this away from you. Please indicate next time that you started working on a fix, such contributions are much appreciated! There are 2596 more issues to look into:)

**Comment by alwaysmpe** [comment]
Haha no worries it's 3am here so I'm not totally coherent.
